### ISSUE TITLE
Handling of `timerange` depends on number of given digits

### ISSUE THREAD

**Issue by schlunma** [ESMValGroup/ESMValCore#2048]
Our handling of the `timerange` option is inconsistent w.r.t. the number of digits given. Examples:

- `timerange: 2000/2001` (or `start_year: 2000`, `end_year: 2001`): extracts the entire years 2000 and 2001 (internally, uses limits 2000-01-01 to 2002-01-01).
- `timerange: 200001/200002`: extracts the entire months 2000-01 and 2000-02 (internally, uses limits 2000-01-01 to 2000-03-01).
- `timerange: 20000101/20000102`: extracts the single day 2000-01-01 (internally, uses limits 2000-01-01 00:00:00 to 2000-01-02 00:00:01).
- Time ranges with even more digits behave like the previous case.

As you can see, in the first two examples, the end point is fully included; in the last one it is **not**.

@ESMValGroup/esmvaltool-coreteam is this something we want? Or is this a weird side effect of our legacy handling of `start_year/end_year`? In any case, I think we need to address this in our documentation [here](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/recipe/overview.html#time-ranges).

Relevant code:

https://github.com/ESMValGroup/ESMValCore/blob/e8efcd863abf0f20c9b08f031d0d5157ea316673/esmvalcore/preprocessor/_time.py#L109-L130

**Comment by bouweandela** [comment]
In most software, (time) ranges are given as [start, stop), i.e. including the start point and excluding the end point from the interval. This has some nice properties:
- the points defining the interval have the same meaning regardless of whether they are used as start or end point. For example: `2000` means `2000-01-01 00:00:00` both when used as a start and as an end point. Compare that with our the current interpretation, where `2000` means `2000-01-01 00:00:00` if used as a start point, but 'anything before 2001-01-01 00:00:00` when used as an end point.
- it is easy to define multiple short time ranges that cover a longer time range because the start point of the next time range can be the same as the endpoint of the previous time range and there will be no gaps
- the is no need to know how many days there are in a month to select the entire month, e.g. XXXX01-XXXX04 will select the first 3 months of the year regardless of which XXXX year is. This makes it a lot easier to handle leap years

The issue with the approach outlined above is that it may not be perceived as intuitive by our community of researchers. That is why we opted for adding 1 to `end_year` when reading it from the recipe. It seems that this approach was only partly implemented for `timerange` when that was implemented to replace `start_year` and `end_year`.

My preference would be to adopt a convention where the endpoint is excluded from the range for all the reasons mentioned above, but that would be an utterly backward incompatible change. Maybe that is something we could consider for a v3 release?

**Comment by katjaweigel** [comment]
Yes, including the end time is often confusing (e.g. that 1990-2000 time ranges in the ESMValTool are 11 years). However, it would be really not good, if existing recipes suddenly produce shorter time series. So if that should be changed, it must be made sure that it is changed in all recipes where it is necessary. And it should be very well communicated, because it means if people use recipes, which are not part of the main the results differ for the same recipe if another ESMValTool version is used.

**Comment by katjaweigel** [comment]
Maybe it could be change through a keyword, e.g. like "endpoint" for numpy.linspace?

**Comment by schlunma** [comment]
> My preference would be to adopt a convention where the endpoint is excluded from the range for all the reasons mentioned above, but that would be an utterly backward incompatible change. Maybe that is something we could consider for a v3 release?

Fully agree! Let's do keep that in mind for v3!

In the meantime, I think we should document this properly. I'll try to come up with a PR soon.
